ttnn/cpp/ttnn/operations/ccl/erisc_datamover_builder.cpp

```diff
@@ -0,0 +1,414 @@
+
+    this->downstream_edm_noc_x = adapter_spec.edm_noc_x;
+    this->downstream_edm_noc_y = adapter_spec.edm_noc_y;
+    this->downstream_edm_buffer_base_address = adapter_spec.edm_buffer_base_addr;
+    this->downstream_edm_semaphore_address = adapter_spec.edm_l1_sem_addr;
+    this->downstream_edm_worker_registration_address = adapter_spec.edm_connection_handshake_addr;
+    this->downstream_edm_worker_location_info_address = adapter_spec.edm_worker_location_info_addr;
+    this->downstream_sender_channel_buffer_index_semaphore_id = adapter_spec.buffer_index_semaphore_id;
+}
+
+
+
+EdmLineFabricOpInterface::EdmLineFabricOpInterface (std::vector<Device*> const& device_sequence, std::vector<Program*> const& program_sequence, std::optional<size_t> desired_num_links) :
+    device_sequence(device_sequence),
+    programs(program_sequence) {
+    static constexpr std::size_t edm_buffer_size = 4096 + sizeof(tt::fabric::PacketHeader);
+    auto const config = FabricEriscDatamoverConfig(edm_buffer_size, 1, 2);
+    TT_ASSERT(device_sequence.size() == program_sequence.size());
+
+    for (size_t i = 0; i < device_sequence.size(); i++) {
+        log_info(tt::LogOp, "device[{}] id={}",  i, device_sequence[i]->id());
+    }
+
+
+    // Construct the builders
+    for (size_t hop = 0; hop < device_sequence.size() - 1; hop++) {
+        auto src_device = device_sequence[hop];
+        auto dest_device = device_sequence[hop + 1];
+
+        auto const& src_device_sockets = src_device->get_ethernet_sockets(dest_device->id());;
+        auto const& dest_device_sockets = dest_device->get_ethernet_sockets(src_device->id());;
+        std::vector<CoreCoord> local_link_cores; local_link_cores.reserve(src_device_sockets.size());
+        std::vector<CoreCoord> remote_link_cores; remote_link_cores.reserve(dest_device_sockets.size());
+        std::copy_if(src_device_sockets.begin(), src_device_sockets.end(), std::back_inserter(local_link_cores), [src_device](CoreCoord const& core) { return src_device->is_active_ethernet_core(core, true); });
+        std::copy_if(dest_device_sockets.begin(), dest_device_sockets.end(), std::back_inserter(remote_link_cores), [dest_device](CoreCoord const& core) { return dest_device->is_active_ethernet_core(core, true); });
+
+        this->num_links = std::min(desired_num_links.value_or(std::numeric_limits<std::size_t>::max()), local_link_cores.size());
+
+        TT_ASSERT(local_link_cores.size() == remote_link_cores.size());
+
+        edm_builders_forward_direction[src_device->id()].reserve(local_link_cores.size());
+        edm_builders_forward_direction[dest_device->id()].reserve(local_link_cores.size());
+        for (size_t l = 0; l < this->num_links; l++) {
+            log_info(tt::LogOp, "Building forward direction EDM on chip {} on link {}", src_device->id(), edm_builders_forward_direction[src_device->id()].size());
```

Comment:
honestly, CCL might deserve its own log category

Reply:
This should actually be a trace... thanks for the reminder, I didn't do my usual info -> trace conversion. Buut yes, I think I should add several categories :)